src/libraries/System.Private.CoreLib/src/System/Reflection/InvokerEmitUtil.cs

```diff
@@ -0,0 +1,191 @@
+                {
+                    il.Emit(OpCodes.Unbox, method.DeclaringType);
+                }
+            }
+
+            // Push the arguments.
+            ParameterInfo[] parameters = method.GetParametersNoCopy();
+            for (int i = 0; i < parameters.Length; i++)
+            {
+                RuntimeType parameterType = (RuntimeType)parameters[i].ParameterType;
+                if (parameterType.IsPointer)
+                {
+                    parameterType = (RuntimeType)typeof(IntPtr);
+                }
+                else if (parameterType.IsByRef && RuntimeTypeHandle.GetElementType(parameterType).IsPointer)
+                {
+                    parameterType = (RuntimeType)typeof(IntPtr).MakeByRefType();
+                }
+
+                il.Emit(OpCodes.Ldarg_2);
+                if (i != 0)
+                {
+                    il.Emit(OpCodes.Ldc_I4, i * IntPtr.Size);
+                    il.Emit(OpCodes.Add);
+                }
+
+                il.Emit(OpCodes.Call, Methods.ByReferenceOfByte_Value()); // This can be replaced by ldfld once byref fields are available in C#
+                if (!parameterType.IsByRef)
+                {
+                    il.Emit(OpCodes.Ldobj, parameterType);
+                }
+            }
+
+            // Invoke the method.
+            if (emitNew)
+            {
+                il.Emit(OpCodes.Newobj, (ConstructorInfo)method);
+            }
+            else if (method.IsStatic || method.DeclaringType!.IsValueType)
+            {
+                il.Emit(OpCodes.Call, (MethodInfo)method);
+            }
+            else
+            {
+                il.Emit(OpCodes.Callvirt, (MethodInfo)method);
+            }
+
+            // Handle the return.
+            if (emitNew)
+            {
+                Type returnType = method.DeclaringType!;
+                if (returnType.IsValueType)
+                {
+                    il.Emit(OpCodes.Box, returnType);
+                }
+            }
+            else
+            {
+                Type returnType = ((RuntimeMethodInfo)method).ReturnType;
+                if (returnType == typeof(void))
+                {
+                    il.Emit(OpCodes.Ldnull);
+                }
+                else if (returnType.IsValueType)
+                {
+                    il.Emit(OpCodes.Box, returnType);
+                }
+                else if (returnType.IsPointer)
+                {
+                    il.Emit(OpCodes.Ldtoken, returnType);
+                    il.Emit(OpCodes.Call, Methods.Type_GetTypeFromHandle());
+                    il.Emit(OpCodes.Call, Methods.Pointer_Box());
+                }
+                else if (returnType.IsByRef)
+                {
+                    // Check for null ref return.
+                    Type elementType = returnType.GetElementType()!;
+                    Label retValueOk = il.DefineLabel();
+                    il.Emit(OpCodes.Dup);
```

Comment:
I recall that working on the Regex compiler we had to remove almost all use of OpCodes.Dup as it was slow. Does that matter here?

Reply:
Single `Dup` in a method is fine. The problem is when you have many `Dup`s in a single method. Every dup is potentially introducing a hidden local variable. One extra local is fine. Hundred of extra locals is not.

Also, this code path will go away eventually.